typescript/infra/scripts/verify-validator.ts

```diff
@@ -0,0 +1,300 @@
+    if (missingInARow == 10) {
+      missingCheckpoints.length -= 10;
+      invalidCheckpoints = invalidCheckpoints.filter((j) => j < i - 10);
+      extraCheckpoints = extraCheckpoints.filter((j) => j < i - 10);
+      break;
+    }
+
+    const key = `checkpoint_${i}.json`;
+
+    let c: Checkpoint | null;
+    try {
+      const t = await cClient.getS3Obj(key);
+      if (isCheckpoint(t)) {
+        if (t.checkpoint.index != i) {
+          console.log(`${i}: Control index is invalid`, t);
+          process.exit(1);
+        }
+        c = t;
+      } else {
+        console.log(`${i}: Invalid control checkpoint`, t);
+        process.exit(1);
+      }
+    } catch (err) {
+      c = null;
+    }
+
+    let p: Checkpoint;
+    try {
+      const t = await pClient.getS3Obj(key);
+      if (isCheckpoint(t)) {
+        p = t;
+      } else {
+        console.log(`${i}: Invalid prospective checkpoint`, t);
+        invalidCheckpoints.push(i);
+        continue;
+      }
+      if (!c) {
+        extraCheckpoints.push(i);
+      }
+      missingInARow = 0;
+    } catch (err) {
+      missingCheckpoints.push(i);
+      missingInARow++;
+      continue;
+    }
+
+    console.assert(
+      p.checkpoint.index == i,
+      `${i}: checkpoint indexes do not match`,
+    );
+
+    // TODO: verify signature
+
+    if (!c) {
+      continue;
+    }
+
+    // compare against the control
+    console.assert(
+      c.checkpoint.outbox_domain == p.checkpoint.outbox_domain,
+      `${i}: outbox_domains do not match`,
+    );
+    console.assert(
+      c.checkpoint.root == p.checkpoint.root,
+      `${i}: checkpoint roots do not match`,
+    );
+
+    try {
+      const [cLastMod, pLastMod] = await Promise.all([
+        cClient.getLastModTime(key),
+        pClient.getLastModTime(key),
+      ]);
+      const diffS = (pLastMod.valueOf() - cLastMod.valueOf()) / 1000;
+      if (Math.abs(diffS) > 10) {
+        console.log(`${i}: Modification times differ by ${diffS}s`);
+      }
+      modTimeDeltasS.push(diffS);
+    } catch (err) {
+      // this is probably a permission error since we already know they should exist
+      console.error(`${i}: Error validating last modified times`, err);
+    }
+
+    fullyCorrectCheckpoints.push(i);
+  }
+
+  if (fullyCorrectCheckpoints.length)
+    console.log(
+      `Fully correct checkpoints (${fullyCorrectCheckpoints.length}): ${fullyCorrectCheckpoints}\n`,
+    );
+  if (extraCheckpoints.length)
+    console.log(
+      `Extra checkpoints (${extraCheckpoints.length}): ${extraCheckpoints}\n`,
+    );
+  if (missingCheckpoints.length)
+    console.log(
+      `Missing checkpoints (${missingCheckpoints.length}): ${missingCheckpoints}\n`,
+    );
+  if (invalidCheckpoints.length)
+    console.log(
+      `Invalid checkpoints (${invalidCheckpoints.length}): ${invalidCheckpoints}\n`,
+    );
+
+  if (modTimeDeltasS.length) {
+    console.log(modTimeDeltasS);
+    console.log(
```

Comment:
nit: might make sense to print this before the one directly above, when running the script it took me a sec to realize the logged array contained the time deltas